Add FilterSet best practices to developer docs (#1691)
nautobot/nautobot: nautobot/docs/development/best-practices.md

```python
class UserFilter(NautobotFilterSet):
    class Meta:
        model = User
        fields = {
            'username': ['exact', 'contains'],
            'last_login': ['exact', 'year__gt'],
        }
```

- It is acceptable that default filter mappings **may** need to be overridden with custom filter declarations, but [`filter_overrides`](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#customise-filter-generation-with-filter-overrides) (see below) should be used as a first resort.
- Custom filter definitions **must not** shadow the name of an existing model field if it is also changing the type.
    - For example `DeviceFilterSet.interfaces` is a `BooleanFilter` that is shadowing the `Device.interfaces` related manager. This introduces problems with automatic introspection of the filterset and this pattern MUST be avoided.

> - Custom filter definitions **may** shadow the name of an existing field IF this is to adapt a filter for a human-readable field value (such as `slug`) on a related field vs. the `pk` (the default for most related fields). This pattern still relies on accessing the related field by name and then traversing into the relationship using a nested lookup.

Reviewer: Should we take this a step further and assert the existing pattern more forcefully? Something like:

> For foreign-key related fields, the field **should** be shadowed, replacing the PK filter with a lookup based on a more human-readable value (typically `slug`, if available). A PK-based filter **should** be made available as well, generally with a name suffixed by `_id`. For example:

```python
    provider = django_filters.ModelMultipleChoiceFilter(
        field_name="provider__slug",
        queryset=Provider.objects.all(),
        to_field_name="slug",
        label="Provider (slug)",
    )
     provider_id = django_filters.ModelMultipleChoiceFilter(
        queryset=Provider.objects.all(),
        label="Provider (ID)",
    )
```

Author: This is tricky because it becomes hard to automate filter generation and/or widgets for this pattern unless we have some other hint to which slug value, like it would automatically always look for the slug on a related field?

Otherwise, by default, `provider` would be a `pk` (UUID) field, whereas if we thought about how we might implement automatic nested lookup expressions, `provider__slug` and  `provider__id` could both "just work". Perhaps we should keep the "should" and the "how" delineated but I feel like this is the time to suss these things out.

Author: Okay unless we go the distance and also implement the primitives I've suggested like our own `SlugPkMultipleChoiceFilter` like some of my prior work:

https://github.com/jathanism/nsot/blob/cecfda0590f2f83f3f679ba415a1d8bdad091238/nsot/api/serializers.py#L85-L139

Then we have to stick with the status quo as well. So is this the right time to assert that this yet-to-be-implemented-but-we-like-the-idea pattern should be used here, and then use that to justify spending the time to implement it?

Reviewer: That's pretty much the question, yes. If there's a new direction we want to go, we can at least insist that any *new* FilterSets (such as Location, say?) follow the new direction rather than propagating an existing bad pattern any further. Can even add a note if you like to the effect of "We know most existing FilterSets aren't following this pattern, but we plan to change that in a major release".

### Filter Naming and Definition

- Boolean filters for membership **must** be named with `has_{related_name}` (e.g. `has_interfaces`)

- Boolean filters for identity **must** be named with `is_{name}` (e.g. `is_virtual_chassis`) although this is semantically identical to `has_` filters, there may be occasions where naming the filter `is_` would be more intuitive.

- Filters **must** declare [`field_name`](https://django-filter.readthedocs.io/en/stable/ref/filters.html#field-name) when they have a different name than the underlying field they are referencing. Where possible the suffix component of the filter name should map directly to the underlying field name.

>   For example, `DeviceFilterSet.has_console_ports` could be better named, to assert that the filter name following the `has_` prefix is a one-to-one mapping to the underlying model's related field name (`consoleports`) which also means that `field_name` is not required:

Reviewer: It's not clear to me why `field_name` would not be required for `has_consoleports`, since it *is* a different name than the underlying `consoleports` field - can you elaborate?

Author: You're right. I done confused myself.

```

- Filters **must** be declared using [`exclude=True`](https://django-filter.readthedocs.io/en/stable/ref/filters.html#exclude) if a queryset `.exclude()` is required to be called vs. queryset `.filter()` which is the default when the filter default `exclude=False` is passed through. If you require `Foo.objects.exclude()`, you must pass `exclude=True` instead of defining a filterset method to explicitly hard-code such a query. For example:

```python
   has_consoleports = BooleanFilter(lookup_expr="isnull", exclude=True)
```

- Filters **must** be declared using [`disinct=True`](https://django-filter.readthedocs.io/en/stable/ref/filters.html#distinct) if a queryset `.distinct()`is required to be called on the queryset
- Filters **must not** be set to be required using `required=True`
- Filter methods defined using the [`method=`](https://django-filter.readthedocs.io/en/stable/ref/filters.html#method) keyword argument **may only be used as a last resort** (see below) when correct usage of `lookup_expr` ,  `exclude`, or other filter keyword arguments do not suffice. In other words: filter methods should used as the exception and not the rule.

> - Use of [`filter_overrides`](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#filter-overrides) **must be considered** in cases where local overrides to certain model field types are required vs. overloading or re-declaring the filter fields are required, such as changing a filter class, or customizing a UI widget. For example:

Reviewer: I'm having a hard time parsing this sentence and understanding when it would apply. When would a "local override to certain model field types" be required? When would "overloading or re-declaring the filter fields" be required?

Author: Refactored.

```python
class ProductFilter(NautobotFilterSet):

     class Meta:
         model = Product
         fields = "__all__"
         filter_overrides = {
             models.CharField: {
                 'filter_class': django_filters.CharFilter,
                 'extra': lambda f: {
                     'lookup_expr': 'icontains',

>                  },

Reviewer: Why a lambda here? Can it not just be `"extra": {"lookup_expr": "icontains"}`? Same question below.

Author: This is how it's implemented in django-filter example documentation. Extra gets passed through as `**kwargs` to the filter on instantiation so I assume it's so that logic can be performed in the callable prior to being passed to the filter on init.

Author: https://django-filter.readthedocs.io/en/stable/ref/filters.html#kwargs

             },
             models.BooleanField: {
                 'filter_class': django_filters.BooleanFilter,
                 'extra': lambda f: {
                     'widget': forms.CheckboxInput,
                 },

>              },

Reviewer: I know this is just an example, but aren't these the defaults for a BooleanField already, so why override them? Can we come up with a meatier example showing a case where this sort of thing is actually necessary?

Author: This was just lifted directly rom the django-filter docs as well, so yes, we could come up with a more specific example.

         }
```

!!!warning
    Existing features of filtersets and filters **must** be exhausted first using keyword arguments before resorting to customizing, re-declaring/overloading, or defining filter methods.

### Filter Methods

> Filter methods in the current Nautobot core are problematic because they break the ability for such filter fields to be properly reversible. Specifically a filter method is a callable or the name of a method on a filterset that can be declared on a filter when it is defined on a filterset. This method is used to perform custom business logic on a filter field.

Reviewer: ```suggestion
Filters on a filterset can reference a `method` (either a callable, or the name of a method on the filterset) to perform custom business logic for that filter field. However, many uses of filter methods in Nautobot are problematic because they break the ability for such filter fields to be properly reversible. 
```

Consider this example from `nautobot.dcim.filters.DeviceFilterSet`:

```python
  # Filter field definition is a BooleanFilter, for which an "isnull" lookup_expr 
  # is the only valid filter expression
  console_ports = django_filters.BooleanFilter(
      method="_console_ports",
      label="Has console ports",
  )
  
  # Method definition loses context and further the field's lookup_expr 
  # falls back to the default of "exact".
  def _console_ports(self, queryset, name, value):
      breakpoint()  # This is where we'll illustrate pdb below
      return queryset.exclude(consoleports__isnull=value)
```

The default `lookup_expr` unless otherwise specified is “exact”, as seen in [django_filters.conf](https://github.com/carltongibson/django-filter/blob/main/django_filters/conf.py#L10):

```python
  'DEFAULT_LOOKUP_EXPR': 'exact',
```

When this method is called, the internal state is default, making reverse introspection impossible, because the `lookup_expr` is defaulting to “exact”:

```python

> (Pdb) field = self.filters[name]

Reviewer 2: ```suggestion
(Pdb) field = self.filters["console_ports"]
```

(Pdb) field.exclude
False
(Pdb) field.lookup_expr
'exact' 

> ```

Reviewer: Is this really an example of why `method`s are bad, or just an example of how core is not currently following all of the requirements laid out above regarding the correct use of `lookup_expr` and `exclude=True`? In other words, if we had instead:

```python
  console_ports = django_filters.BooleanFilter(
      method="_console_ports",
      label="Has console ports",
      exclude=True,
      lookup_expr="isnull",
  )
```

the method would still be *redundant* (as laid out below), but would it still be *a problem* as far as introspection and reversibility?

Or to put it yet another way, do we have any better examples where we have filter fields that are properly defined per all of the above requirements, but still need to use `method`, and that usage is still a problem?

Author: Solid point. I think we could refresh this to use a different field. I'll come up with one.

Author: To answer your questions, however:

> the method would still be redundant (as laid out below), but would it still be a problem as far as introspection and reversibility?

Redundant, yes. Problematic, maybe, since it provides context on how to proceed, if it's doing exactly the same logic as what the method would be doing, then it's not a problem. It becomes more problematic in the case where we don't actually know *what* the method is doing internally, and for example if it's generating a `Q` object internally. As developers/authors, that's not bad, as platfors maintainers trying to build reversible lookups, that's bad.

Author: For the sake of this thread, I would argue that `DeviceFilterSet.pass_through_ports` is bad. 

Here's why:

1. It's a method filter:

```python
    pass_through_ports = django_filters.BooleanFilter(
        method="_pass_through_ports",
        label="Has pass-through ports",
    )
```

2. It's generating a compound exclude filter using two related fields:

```python
    # 2.0 TODO: Remove me and `pass_through_ports` in exchange for `has_(front|rear)_ports`.
    def _pass_through_ports(self, queryset, name, value):
        return queryset.exclude(frontports__isnull=value, rearports__isnull=value)
```

3. `field_name` is irrelevant here because there are two at play (`frontports` and `rearports`)

4. There is no way to introspect what is going on in the method body with 

So while this filter definition could be improved like so:

```python
    pass_through_ports = django_filters.BooleanFilter(
        method="_pass_through_ports",
        label="Has pass-through ports",
        lookup_expr="isnull",
        exclude=True,
    )
```

It stops there.

- There's no way to identify either of the field names required here
- The `name` that is incoming to the method is the filter name as defined `has_pass_through_ports` in this case
- So the filter can be introspected for `lookup_expr` value using `self.filters[name].lookup_expr`, but it would have to be assumed that applies to both.
- Same with `exclude` (`self.filters[name].exclude`)

It would be better to just eliminate `pass_through_ports=True` entirely in exchange for `front_ports=True&rear_ports=True`  (current) or `has_frontports=True&has_rearports=True` (future)

This means that the arguments for the field are being completely ignored and the hard-coded queryset `queryset.exclude(consoleports__isnull=value)` is all that is being run when this method is called. This hard-coding is impossible to introspect and therefore reverse.

In fact, this method is completely unnecessary, because it could be replaced with options on the filter field itself:

```python
    console_ports = django_filters.BooleanFilter(
        field_name="consoleports",  # The actual related field name
        exclude=True,               # Perform an `.exclude()` vs. `.filter()``
        lookup_expr="isnull",       # Perform `isnull` vs. `exact``
        label="Has console ports",
    )
```

> Now, if we use another breakpoint (this time inside of the `DeviceListView.get()` since there’s no method we can break into), you can see that the filter field now has the correect attributes that can be used to accurately reverse this query:

Reviewer: ```suggestion
Now, if we use another breakpoint (this time inside of the `DeviceListView.get()` since there’s no method we can break into), you can see that the filter field now has the correct attributes that can be used to accurately reverse this query:
```

```python
(Pdb) filterset = self.filterset(request.GET, self.queryset)
(Pdb) field = filterset.filters["console_ports"]
(Pdb) field.exclude
True
(Pdb) field.lookup_expr
'isnull'
```

#### Generating Reversible Q Objects

These field values could be used to construct a `Q()` query that looks something like:

```python
def generate_query(self, field_name, lookup_expr, value):
    query = Q()
    predicate = {f"{field_name}__{lookup_expr}": value}
    if field.exclude:
        query |= ~Q(**predicate)
    else:
        query |= Q(**predicate)
    return query
    
## Somewhere else in business logic:
field = filterset.filters[name]
value = filterset.data[name]
query = generate_query(field.field_name, field.lookup_expr, value)
filterset.qs.filter(query).count()  # 339
```

But furthermore, this also becomes unnecessary for the common case, because we can **just trust the filterset itself to do the proper filtering and completely obviate the need to reverse the query at all.**

Consider this input filter:

```python
filter_params = {"console_ports": True}
```

With properly configured filter fields, this will just work:

```python
DeviceFilterSet(filter_params).qs.count()  # 339

> ```

Reviewer: Not sure I understand what this section is arguing for or against, or guiding me to do or not do as far as coding best practices. It seems to be saying "Hey, you could do this, but it would be redundant" - so what's the takeaway?

Author: This entire section of the document could probably just be left out or replaced since it was a part of my initial analysis/discovery. 

On that note, it almost feels like my [response to comment in above](https://github.com/nautobot/nautobot/pull/1691#discussion_r863167001) could be what replaces it. Thoughts there?

The other thing that comes to mind is more of a "what if" (similarly to the slug/id filter field) we implemented a "composite" field that took arguments of the various field names (such as in the case of `DeviceFilterSet.pass_through_ports` that could facilitate such composite lookups but still provide arguments/attributes that could be use to inform query construction.

### Summary

- For the vast majority of cases where we have method filters, it’s for Boolean filters
- I am asserting for the common case **method filters are unnecessary technical debt and should be eliminated where better suited by proper use of filter field arguments**
- Reversibility may not necessarily be required, but by properly defining `field_name`, `lookup_expr`, and `exclude` on filter fields, **introspection becomes deterministic and reversible queries can be reliably generated as needed.**

> - For exceptions such as `DeviceFilterSet.has_primary_ip` where it checks for both `Device.primary_ip4` OR `Device.primary_ip6`, method filters may still be necessary, however, they would be **the exception and not the norm.**

Reviewer: I'd love to see this one fleshed out more - here we have a filter field that *does* need to use a `method` - so is it currently implemented as optimally and correctly as possible? Does it still pose problems, and if so what are they?

Author: This seems to also be answered by the same comment above. However, could still refine this into the pros/cons of using method filters and best practices for IF you do vs. why you shouldn't or try to avoid it?

- For foreign-key related fields, **on existing core models in the v1.3 release train**: 
    - The field **should** be shadowed, replacing the PK filter with a lookup-based on a more human-readable value (typically `slug`, if available). 
    - A PK-based filter **should** be made available as well, generally with a name suffixed by `_id`. For example:

```python
    provider = django_filters.ModelMultipleChoiceFilter(
        field_name="provider__slug",
        queryset=Provider.objects.all(),
        to_field_name="slug",
        label="Provider (slug)",
    )
    provider_id = django_filters.ModelMultipleChoiceFilter(
        queryset=Provider.objects.all(),
        label="Provider (ID)",
    )
```

- For foreign-key related fields on **new core models for v1.4 or later:**
    - The field **must** be shadowed utilizing a hybrid `NaturalKeyMultipleChoiceFilter` which will automatically try to lookup by UUID or `slug` depending on the value of the incoming argument (e.g. UUID vs. slug). 

>     - Using this filter by default, `provider` would be a `pk` (UUID) field, whereas ithis field will automatically support both input values for `slug` or `pk`.

Reviewer: Not sure I parse this sentence. Is "this filter" referring to `NaturalKeyMultipleChoiceFilter` or the default (pk-only) filtering?

    - New filtersets should follow this direction vs. propagating the need to continue to overload `provider` and define an additional `provider_id` filter on each new filterset. *We know that most existing FilterSets aren't following this pattern, and we plan to change that in a major release.*
    - Using the previous field (`provider`) as an example, it would look something like this:

```python
    provider = NaturalKeyModelMultipleChoiceFilter(
        queryset=Provider.objects.all(),

>         to_field_name="slug",

Author: Great question. I'm thinking we have it generate a `Q` object and store it similarly to what we did with `SearchFilter`
